include/preprocess/crhmc/analytic_center.h

```diff
@@ -0,0 +1,178 @@
+//     If the domain ({Ax=b} intersect dom(f)) is not full dimensional in {Ax=b}
+//     because of the dom(f), the algorithm will detect the collapsed dimension
+//     and output the detected constraint C x = d
+// d - detected constraint vector
+template <typename Polytope>
+std::tuple<VT, SpMat, VT> analytic_center(SpMat const &A, VT const &b, Polytope &f, Opts const &options, VT x = VT::Zero(0, 1))
+{
+  // initial conditions
+  int n = A.cols();
+  int m = A.rows();
+  if (x.rows() == 0 || !f.barrier.feasible(x))
+  {
+    x = f.barrier.center;
+  }
+
+  VT lambda = VT::Zero(n, 1);
+  int fullStep = 0;
+  NT tConst = 0;
+  NT primalErr = std::numeric_limits<NT>::infinity();
+  NT dualErr = std::numeric_limits<NT>::infinity();
+  NT primalErrMin = std::numeric_limits<NT>::infinity();
+  NT primalFactor = 1;
+  NT dualFactor = 1 + b.norm();
+  std::vector<int> idx;
+
+  CholObj solver = CholObj(A);
+
+  for (int iter = 0; iter < options.ipmMaxIter; iter++)
+  {
+    std::pair<VT, VT> pair_analytic_oracle = f.analytic_center_oracle(x);
+    VT grad = pair_analytic_oracle.first;
+    VT hess = pair_analytic_oracle.second;
+
+    // compute the residual
+    VT rx = lambda - grad;
+    VT rs = b - A * x;
+
+    // check stagnation
+    NT primalErrMin = std::min(primalErr, primalErrMin);
+    NT primalErr = rx.norm() / primalFactor;
+    NT dualErrLast = dualErr;
+    NT dualErr = rs.norm() / dualFactor;
+    bool feasible = f.barrier.feasible(x);
+    //Compare the dual and primal error to the last and minimum so far
+    if ((dualErr > (1 - 0.9 * tConst) * dualErrLast) ||
+        (primalErr > 10 * primalErrMin) || !feasible)
+    {
+      VT dist = f.barrier.boundary_distance(x);
+      NT th = options.ipmDistanceTol;
+      visit_lambda(dist, [&idx, th](double v, int i, int j)
+                   {
+        if (v < th)
+          idx.push_back(i); });
+      if (idx.size() > 0)
+      {
+        break;
+      }
+    }
+
+    // compute the step direction
+    VT Hinv = hess.cwiseInverse();
+    solver.decompose((Tx *)Hinv.data());
+    VT out(m, 1);
+    solver.solve((Tx *)rs.data(), (Tx *)out.data());
+    VT dr1 = A.transpose() * out;
+    VT in = A * Hinv.cwiseProduct(rx);
+    solver.solve((Tx *)in.data(), (Tx *)out.data());
+
+    VT dr2 = A.transpose() * out;
+    VT dx1 = Hinv.cwiseProduct(dr1);
+    VT dx2 = Hinv.cwiseProduct(rx - dr2);
+
+    // compute the step size
+    VT dx = dx1 + dx2;
+    NT tGrad = std::min(f.barrier.step_size(x, dx), 1.0);
+    dx = dx1 + tGrad * dx2;
+    NT tConst = std::min(0.99 * f.barrier.step_size(x, dx), 1.0);
+    tGrad = tGrad * tConst;
+
+    // make the step
+    x = x + tConst * dx;
+    lambda = lambda - dr2;
+
+    if (!f.barrier.feasible(x))
+    {
+      break;
+    }
+    //If we have have converged
+    if (tGrad == 1)
+    {
+      //do some more fullStep
+      fullStep = fullStep + 1;
+      if (fullStep > log(dualErr / options.ipmDualTol) && fullStep > 8)
```

Comment:
do we know where 8 comes from here?

Reply:
I do not know what is exactly the reasoning but it acts as a minimum step number. I placed it in the options file as a preprocessing constant.